packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -3643,14 +3643,28 @@ class _UpdateTextSelectionAction<T extends DirectionalCaretMovementIntent> exten
       );
     }
 
-    final TextPosition extent = textBoundarySelection.extent;
+    late final TextPosition position;
+    if (!intent.expand) {
+      position = textBoundarySelection.extent;
+    } else {
+      final bool inOrder = textBoundarySelection.baseOffset <= textBoundarySelection.extentOffset;
+      final bool towardsExtent = intent.forward == inOrder;
+      position = towardsExtent
+          ? textBoundarySelection.extent
+          : textBoundarySelection.base;
+    }
     final TextPosition newExtent = intent.forward
-      ? textBoundary.getTrailingTextBoundaryAt(extent)
-      : textBoundary.getLeadingTextBoundaryAt(extent);
+      ? textBoundary.getTrailingTextBoundaryAt(position)
+      : textBoundary.getLeadingTextBoundaryAt(position);
 
-    final TextSelection newSelection = collapseSelection
-      ? TextSelection.fromPosition(newExtent)
-      : textBoundarySelection.extendTo(newExtent);
+    late final TextSelection newSelection;
+    if (collapseSelection) {
+      newSelection = TextSelection.fromPosition(newExtent);
+    } else {
+      newSelection = intent.expand
```

Comment:
I tried this and it seems macos does not move the extent when it expands the selection? cmd+shift+-> and then press -> the cursor moves to 1 character to the right from its original position.

Reply:
I think I've got this right now, the behavior might have changed since you tested it.  Here's what I'm testing that seems to match with TextEdit:

  * With cursor in the middle of the line, cmd + left.  Left end is the extent.
  * With cursor in the middle of the line, cmd + right.  Right end is the extent.
  * With cursor in the middle of the line, cmd + left then cmd + right.  Left end is the extent.
  * With cursor in the middle of the line, cmd + right then cmd + left.  Right end is the extent.
  * With an RTL selection in the middle of the line, cmd + left.  Left end is the extent.
  * With an RTL selection in the middle of the line, cmd + right.  Left end is the extent.
  * With an LTR selection in the middle of the line, cmd + right.  Right end is the extent.
  * With an LTR selection in the middle of the line, cmd + left.  Right end is the extent.

Reply:
I just added tests for all of these cases too.  By "testing" in the previous comment I meant manually trying myself vs. TextEdit.